docs/daemon_design.md

```diff
@@ -0,0 +1,685 @@
+
+We will move forward with the second option, *one scheduling manager per environment*, to make scaling and cleanup of managers simpler, and ensure that customers do not affect each other.
+
+The controller will be responsible for creating a scheduling manager for every new environment, making sure that the scheduling manager is alive and healthy, and restarting it if it dies. The controller will keep track of manager health by expecting regular heartbeats from the scheduling manager. If heartbeats are not received, the manager will be restarted.
+
+```
+Table: Scheduling_Managers
+Key: EnvironmentName
+SortKey: ManagerId
+```
+
+```
+Starting a manager
+
+if a new environment is created (listen to dynamodb stream)
+    create manager workflow {
+
+        if cluster does not exist for the environment customer
+            cluster creation workflow {
+                create an autoscaling group
+                create cluster
+                wait for instances in the cluster to become available
+            }
+
+        if (less than desired number of managers exist for the environment OR
+           lastUpdatedTime is older than the expected heartbeat interval) {
+
+            create a new manager record with environmentName and
+                new managerId and status=unregistered 
+        }
+
+        if manager record put successful
+            scheduling manager creation workflow {
+                start a manager task in the customer cluster
+                autoscale the cluster if needed
+                describe task until ready
+
+                if task cannot be started
+                    update environment with status launch_failed
+                else if task is running
+                    update environment to launched
+            }
+    }
+```
+
+If the controller instance dies in the middle of starting a manager, we need to be able to pick up the work where it was left off and continue. To achieve this we need to save state after each operation so the next worker knows where the work was interrupted, i.e. a workflow. We will use AWS Stepfunctions to create a workflow: state handling and retries will be handled by the stepfunctions service.
+
+##### Monitoring and Lifecycle
+
+We might get into situations where there are multiple managers running for an environment, for example, if a manager cannot send a heartbeat but is still processing deployments. The managers will "lock" the deployment when they pick it up for processing by updating the environment record to "being-processed" using dynamodb conditional writes to avoid race conditions. This will ensure that even if there are multiple managers running for an environment, they don't do duplicate work. The manager might die before releasing the lock. We will need a monitor to go through all deployments and release the lock if it's been held for too long. We also need a monitor to sweep for duplicate managers and clean them up.
```

Comment:
Locks and a background monitor to release them can be a bit tough to operate.  I've had good luck following a "distributed lease" model where a worker attempts to increment a version number with a conditional write along with a TTL.  All other workers start a local timer every time they see the version number increase and only try to acquire it if the timer has expired.  There are a couple libraries out there that make this easy, and I think @Yanacek may have had one under aws-labs before?